### PR TITLE
segmentation evaluations

### PR DESCRIPTION
**Pull request by mklarqvist**
compares [1,N) number of segmentation models and returns
* model evaluations (default keras)
* batch-wise model evaluations
* predictions (default keras)
* sparse pixel-wise classifications

can accept `tf.data.Dataset` and `ml4h.tensor_generators.TensorGenerator`

### CODE REVIEW

File: ml4h/arguments.py [hunk 1]
```diff
@@ -336,6 +336,9 @@ def parse_args():
     # TensorMap prefix for convenience
     parser.add_argument('--tensormap_prefix', default="ml4h.tensormap", type=str, help="Module prefix path for TensorMaps. Defaults to \"ml4h.tensormap\"")
 
+    # Special
+    parser.add_argument('--is_segmentation', default=False, help='Trigger segmentation-specific functionality.')
```
Comment (lucidtronix) [inline ml4h/arguments.py:340]: I think we can remove this and use tensormap interpretation and shape check instead

Comment (lucidtronix) [reply ml4h/arguments.py:340]: see below

File: ml4h/recipes.py [hunk 2]
```diff
@@ -175,6 +178,12 @@ def compare_multimodal_multitask_models(args):
     input_data, output_data, paths = big_batch_from_minibatch_generator(generate_test, args.test_steps)
     common_outputs = _get_common_outputs(models_inputs_outputs, 'output')
     predictions = _get_predictions(args, models_inputs_outputs, input_data, common_outputs, 'input', 'output')
+    if args.is_segmentation:
```
Comment (lucidtronix) [inline ml4h/recipes.py:181]: i think we can do `if any([tm.is_categorical and tm.axes() > 1 for tm args.tensor_maps_out]):`

File: ml4h/evaluate_segmentations.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        if isinstance(g, dataset_ops.DatasetV2):
+            if hasattr(g, '_batch_size'):
+                if g._batch_size.numpy() != 1:
+                    raise ValueError("Batch size must be set to 1")
+        elif isinstance(g, TensorGenerator):
+            files = np.array([p.paths for p in g.path_iters]).flatten()
+            print(files)
+            g = DataGenerator(files, g)
+        else: # yolo
+            pass
```
Comment (lucidtronix) [inline ml4h/evaluate_segmentations.py:128]: lol, at least log a warning

File: ml4h/evaluate_segmentations.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        # tf.data.Dataset family of generators in TensorFlow 2+
+        # Todo: could eventually be extended to include checks for 
+        # `dataset._dataset._batch_size` that is used in TensorFlow 1.*
+        if isinstance(g, dataset_ops.DatasetV2):
+            if hasattr(g, '_batch_size'):
+                if g._batch_size.numpy() != 1:
+                    raise ValueError("Batch size must be set to 1")
+        elif isinstance(g, TensorGenerator):
+            files = np.array([p.paths for p in g.path_iters]).flatten()
+            print(files)
```
Comment (lucidtronix) [inline ml4h/evaluate_segmentations.py:125]: do we need this print?

File: ml4h/evaluate_segmentations.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+# Given a TensorFlow model, a dictionary of metrics, and a data generator
+# or (x,y) pair, we evaluate the model and return a pair of dictionary for
+# per-batch (n=1) metrics and overall mean metrics.
+# 
+# As a preprocessing step the input `model` must already be re/compiled
+# to use the exact metrics in the `metrics` list.
+def evaluate_collect_metrics(model: tf.keras.models.Model, 
+        data_generator: list, 
+        metrics: List[Callable] = None):
```
Comment (lucidtronix) [inline ml4h/evaluate_segmentations.py:61]: The type annotation says List, but it seems you pass in a dict?

File: ml4h/evaluate_segmentations.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    #
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        metrics_dict = {m.name: m for m in metrics}
+    #
+    model_count = 0
+    evaluations = list()
+    for m,g in zip(models, data_generators):
+        if metrics is None:
```
Comment (lucidtronix) [inline ml4h/evaluate_segmentations.py:113]: What is the difference between what happens in line 107 and what happens here

File: ml4h/recipes.py [hunk 7]
```diff
@@ -207,6 +216,79 @@ def inference_file_name(output_folder: str, id_: str) -> str:
...
+
+    Evaluations returned are:
+        * Per-model metrics
+        * Per-input metrics (batch-size is 1)
+        * The soft model predictions
+        * The hard model predictions
+    """
+    # Local includes only required in this case.
+    import ml4h.evaluate_segmentations as evaluate
+    from ml4h.tensor_generators import TensorGenerator
```
Comment (lucidtronix) [inline ml4h/recipes.py:234]: this is already included at top

File: ml4h/recipes.py [hunk 8]
```diff
@@ -629,6 +711,53 @@ def _get_predictions(args, models_inputs_outputs, input_data, outputs, input_pre
...
+                    },
+                'tensor_map_2':
+                    {
+                        'model_2': [[value5, value6], [...]],
+                        'model_3': [[value7, value8], [...]]
+                    }
+            }
+    """
+    # Local includes only required in this case.
+    import ml4h.evaluate_segmentations as evaluate
```
Comment (lucidtronix) [inline ml4h/recipes.py:734]: For consistency, use file level import and prefer explicit imports without aliasing when possible, so `from ml4h.evaluate_segmentations import evaluate_segmentation_models`

File: ml4h/recipes.py [hunk 9]
```diff
@@ -629,6 +711,53 @@ def _get_predictions(args, models_inputs_outputs, input_data, outputs, input_pre
...
+    from ml4h.tensor_generators import TensorGenerator
```
Comment (lucidtronix) [inline ml4h/recipes.py:735]: already imported

File: ml4h/recipes.py [hunk 10]
```diff
@@ -629,6 +711,53 @@ def _get_predictions(args, models_inputs_outputs, input_data, outputs, input_pre
...
+            args.tensor_maps_out,
+            paths=input_data,
+            num_workers=1,
+            cache_size=0, 
+            keep_paths=True, 
+            mixup=args.mixup_alpha,
+        )
+
+        eval = evaluate.evaluate_segmentation_models([model], [generate_test])
+        logging.info(f"Evaluations completed.")
```
Comment (lucidtronix) [inline ml4h/recipes.py:756]: `logging.info(f"Evaluations completed for model: {model_file}.")`

File: ml4h/test_evaluate.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import tensorflow as tf
```
Comment (lucidtronix) [inline ml4h/test_evaluate.py:1]: Can you do this via pytest in tests dir?

**Review by lucidtronix — CHANGES_REQUESTED** [review]
@mklarqvist Great work! Excited to use this-- a couple refactor requests and a couple questions.

File: ml4h/test_evaluate.py (hunk 11 above)
Comment (mklarqvist) [reply ml4h/test_evaluate.py:1]: Ah yes, we can delete this file. This was my local debugging file

File: ml4h/evaluate_segmentations.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        metrics: List[Callable] = None):
+    # Reformat metrics list to a metrics dictionary of the form
+    # {name: function pointer}.
+    if metrics is not None:
+        if isinstance(metrics, list):
+            metrics_dict = {m.name: m for m in metrics}
+        else:
+            metrics_dict = metrics
+    else:
+        metrics_dict = {m.name: m for m in model.metrics}
```
Comment (mklarqvist) [inline ml4h/evaluate_segmentations.py:70]: It can be either a List or a Dict. I'll remove the type hint

File: ml4h/evaluate_segmentations.py (hunk 6 above)
Comment (mklarqvist) [reply ml4h/evaluate_segmentations.py:113]: It should be such that 1) if `metrics` is passed then used that for all models, otherwise 2) foreach model use whatever metrics it came with (`model.metrics`)

File: ml4h/evaluate_segmentations.py (hunk 3 above)
Comment (mklarqvist) [reply ml4h/evaluate_segmentations.py:121]: This should throw an exception and captured by Keras/Tensorflow later if the generator is not legal

**Comment by mklarqvist** [comment]
Now also works through the command line through the `evaluate_batch` recipe. It cannot support most of the functionality in the standalone callback function. Example:
```bash
python3 /tf/ml/ml4h/recipes.py \
--mode evaluate_batch \
--tensors /mnt/disks/annotated-cardiac-tensors-45k/2020-09-21/ \
--tensormap_prefix ml4h.tensormap.ukb.mri \
--model_file sax_slices_jamesp_test_no_bn_v2.h5 \
--input_tensors sax_slice_jamesp \
--output_tensors cine_segmented_sax_no_pap_slice_jamesp \
--output_folder /tf/eval_outputs/ \
--id sax_slices_jamesp_test_no_bn_v2 \
--sample_csv /tf/sax_test.csv

```

**Comment by mklarqvist** [comment]
The CLI code path will not provide the following outputs:

* model evaluations (default keras) --- averaging the batch outputs will generally not produce the correct output
* predictions (default keras) --- use an inference recipe
* sparse pixel-wise classifications --- not supported at the moment

**Comment by mklarqvist** [comment]
Now  includes static image generation for each metric that was compiled with the trained model
sax_slices_jamesp_4b_hyperopted_dropout_pap_dupe__test_data__model0__inference_evaluate_batch_summary_stats_RV_cavity_precision
sax_slices_jamesp_4b_hyperopted_dropout_pap_dupe__test_data__model0__inference_evaluate_batch_summary_stats_RV_free_wall_precision
sax_slices_jamesp_4b_hyperopted_dropout_pap_dupe__test_data__model0__inference_evaluate_batch_summary_stats_RV_cavity_recall

sax_slices_jamesp_4b_hyperopted_dropout_pap_dupe__test_data__model0__inference_evaluate_batch_summary_stats_RV_free_wall_recall

File: ml4h/evaluate_batch.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    def __getitem__(self, index):
+        x, y, _, b = next(self.generator)
+        self.offset += 1
+        return x, y, [None]
+    
+    def on_epoch_end(self):
+        self.offset = 0
+
+
+# Given a TensorFlow model, a dictionary of metrics, and a data generator
```
Comment (lucidtronix) [inline ml4h/evaluate_batch.py:47]: Make multi-line comment inside function definition

File: ml4h/evaluate_batch.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    eval = {out: eval[i] for i, out in enumerate(model.metrics_names)}
+    eval = pd.DataFrame(eval,index=[0])
+    return eval, eval_batch#, prediction_prob, predictions
+
+
+def evaluate_models(models: List[tf.keras.models.Model], 
+        data_generators: list, 
+        metrics: List[Callable] = None,
+        channel_order: str = 'channels_last'):
+    # `metrics` must be either a list of callable metrics functions
```
Comment (lucidtronix) [inline ml4h/evaluate_batch.py:89]: or a Dict?

File: ml4h/recipes.py [hunk 15]
```diff
@@ -207,6 +217,123 @@ def inference_file_name(output_folder: str, id_: str) -> str:
...
+    """Evaluates batch-wise information for a model given the metrics
+    it was compiled with. This temporary function works only with ML4H
+    TensorGenerators.
+    """
+    # Local includes only required in this case.
+    from tensorflow.keras.models import load_model
+    from ml4h.metrics import get_metric_dict
+    # Local includes only required in this case.
+    from ml4h.evaluate_segmentations import BatchMetricsLogger
+    import seaborn as sns
```
Comment (lucidtronix) [inline ml4h/recipes.py:230]: move plotting to plots.py and maybe wait for subsequent PR to add plotting

Comment (lucidtronix) [reply ml4h/recipes.py:230]: What are these plots showing?  What is Rank referring to?

**Review by lucidtronix — CHANGES_REQUESTED** [review]
To merge this sooner rather than later I think it's best to hold off on automated plotting and just make sure the TSVs generated here work for any model produced by the modelfactory.

File: ml4h/recipes.py (hunk 15 above)
Comment (mklarqvist) [reply ml4h/recipes.py:230]: Sort order (rank)

**Review by lucidtronix — CHANGES_REQUESTED** [review]
It looks like the file `evaluate_segmentations` is missing from this commit.
